src/bin/lamp.rs

```diff
@@ -1,228 +1,782 @@
+}
 
-    let on_description = json!({
-        "@type": "OnOffProperty",
-        "title": "On/Off",
-        "type": "boolean",
-        "description": "Whether the lamp is turned on"
-    });
-    let on_description = on_description.as_object().unwrap().clone();
-    thing.add_property(Box::new(BaseProperty::new(
-        "on".to_owned(),
-        json!(true),
-        None,
-        Some(on_description),
-    )));
-
-    let brightness_description = json!({
-        "@type": "BrightnessProperty",
-        "title": "Brightness",
-        "type": "integer",
-        "description": "The level of light from 0-100",
-        "minimum": 0,
-        "maximum": 100,
-        "unit": "percent"
-    });
-    let brightness_description = brightness_description.as_object().unwrap().clone();
-    thing.add_property(Box::new(BaseProperty::new(
-        "brightness".to_owned(),
-        json!(50),
-        None,
-        Some(brightness_description),
-    )));
-
-    let fade_metadata = json!({
-        "title": "Fade",
-        "description": "Fade the lamp to a given level",
-        "input": {
-            "type": "object",
-            "required": [
-                "brightness",
-                "duration"
-            ],
-            "properties": {
-                "brightness": {
-                    "type": "integer",
-                    "minimum": 0,
-                    "maximum": 100,
-                    "unit": "percent"
-                },
-                "duration": {
-                    "type": "integer",
-                    "minimum": 1,
-                    "unit": "milliseconds"
-                }
-            }
+impl Default for ActionStatus {
+    fn default() -> Self {
+        Self::Created
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum StoredActionType {
+    Fade(StoredFadeAction),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct StoredFadeAction {
+    input: FadeActionInput,
+    href: String,
+    #[serde(with = "time::serde::rfc3339")]
+    time_requested: OffsetDateTime,
+    #[serde(
+        with = "time::serde::rfc3339::option",
+        skip_serializing_if = "Option::is_none"
+    )]
+    time_completed: Option<OffsetDateTime>,
+    #[serde(default)]
+    status: ActionStatus,
+}
+
+async fn post_actions(
+    Json(value): Json<Value>,
+    Extension(state): Extension<AppState>,
+) -> Result<impl IntoResponse, AppError> {
+    let action: Action = serde_json::from_value(value).map_err(|_| AppError::InvalidValue)?;
+    match action {
+        Action::Fade(fade) => {
+            let action = state.fade(fade.input).await;
+            let (input, href) = match action.ty {
+                StoredActionType::Fade(StoredFadeAction { input, href, .. }) => (input, href),
+            };
+
+            let response = (
+                StatusCode::CREATED,
+                Json(json!({
+                    "fade": {
```

Comment:
It should be an [ActionStatus](https://w3c.github.io/wot-profile/#ActionStatus) and not start with the named wrapper.

Reply:
I think you are almost surely right... which is quite bad, because it demonstrates that the tests are pretty unreliable...